HW1/P5/P5_bfs.py

```diff
@@ -0,0 +1,102 @@
+from pyspark import AccumulatorParam
+
+# Accumulator used for BFS. Basically the algorithm adds all the visited
+# nodes in one iteration into this accumulator instance.
+class SetAccm(AccumulatorParam):
+    def zero(self, s):
+        return s
+
+    def addInPlace(self, s1, s2):
+        s1 |= s2
+        return s1
+
+def bfs(graph, source):
+    sc = graph.context
+    childParent = sc.parallelize([(source, -1)])
+
+    frontier, closed = set([source]), set([source])
+
+    while True:
+        # Initialize the accumulator in each iteration
+        tmpFoundAccm = sc.accumulator(set(), SetAccm())
+        # Those entries in frontier queue will be searched
+        toSearch = graph.filter(lambda (k, _): k in frontier)
+        # Add all the adjacent nodes into the accumulator
+        toSearch.foreach(lambda (k, adj): tmpFoundAccm.add(set(adj)))
+
+        frontier = tmpFoundAccm.value - closed
```

Comment:
This is done in python, but should have been done in spark (using joins, subtractByKey, and unions).